modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -169,42 +181,84 @@ object Enrich {
+    }.separate
+
+    val allBad = (bad ++ moreBad).map(badRowResize(env, _))
+
+    List(
+      sinkGood(
+        good,
+        env.sinkGood,
+        env.metrics.goodCount,
+        env.metadata.observe,
+        env.sinkPii,
+        env.piiAttributes,
+        env.processor,
+        env.streamsSettings.maxRecordSize
+      ) *> env.metrics.enrichLatency(chunk.headOption.flatMap(_._2)),
+      sinkBad(allBad, env.sinkBad, env.metrics.badCount)
+    ).parSequence_
+  }
 
-  def sinkOne[F[_]: Concurrent: Parallel, A](env: Environment[F, A])(result: Validated[BadRow, EnrichedEvent]): F[Unit] =
-    result.fold(sinkBad(env, _), sinkGood(env, _))
-
-  def sinkBad[F[_]: Monad, A](env: Environment[F, A], bad: BadRow): F[Unit] =
-    env.metrics.badCount >> env.sinkBad(badRowResize(env, bad))
-
-  def sinkGood[F[_]: Concurrent: Parallel, A](env: Environment[F, A], enriched: EnrichedEvent): F[Unit] =
-    serializeEnriched(enriched, env.processor, env.streamsSettings.maxRecordSize) match {
-      case Left(br) => sinkBad(env, br)
-      case Right(bytes) =>
-        List(env.metrics.goodCount,
-             env.metadata.observe(enriched),
-             env.sinkGood(AttributedData(bytes, env.goodAttributes(enriched))),
-             sinkPii(env, enriched)
-        ).parSequence_
-    }
+  def sinkGood[F[_]: Sync](
+    good: List[(EnrichedEvent, AttributedData[Array[Byte]])],
+    sink: AttributedByteSink[F],
+    incMetrics: Int => F[Unit],
+    metadata: List[EnrichedEvent] => F[Unit],
+    piiSink: Option[AttributedByteSink[F]],
+    piiAttributes: EnrichedEvent => Map[String, String],
+    processor: Processor,
+    maxRecordSize: Int
+  ): F[Unit] = {
+    val enriched = good.map(_._1)
+    val serialized = good.map(_._2)
+    sink(serialized) *> incMetrics(good.size) *> metadata(enriched) *> sinkPii(enriched, piiSink, piiAttributes, processor, maxRecordSize)
+  }
 
-  def sinkPii[F[_]: Monad, A](env: Environment[F, A], enriched: EnrichedEvent): F[Unit] =
-    (for {
-      sink <- env.sinkPii
-      pii <- ConversionUtils.getPiiEvent(env.processor, enriched)
-    } yield serializeEnriched(pii, env.processor, env.streamsSettings.maxRecordSize) match {
-      case Left(br) => sinkBad(env, br)
-      case Right(bytes) => sink(AttributedData(bytes, env.piiAttributes(pii)))
-    }).getOrElse(Monad[F].unit)
+  def sinkBad[F[_]: Monad](
+    bad: List[Array[Byte]],
+    sink: ByteSink[F],
+    incMetrics: Int => F[Unit]
+  ): F[Unit] =
+    sink(bad) *> incMetrics(bad.size)
+
+  def sinkPii[F[_]: Sync](
+    enriched: List[EnrichedEvent],
+    maybeSink: Option[AttributedByteSink[F]],
+    attributes: EnrichedEvent => Map[String, String],
+    processor: Processor,
+    maxRecordSize: Int
+  ): F[Unit] =
+    maybeSink match {
+      case Some(sink) =>
+        val (bad, serialized) =
+          enriched
+            .map(ConversionUtils.getPiiEvent(processor, _))
+            .collect { case Some(e) => e }
```

Comment:
Only sharing this because it's kind fun....

I think this can be `.flatMap(ConversionUtils.getPiiEvent(processor, _))`  instead of `.map(....).collect(....)`

Untested though.

Reply:
Yes good point !